text/0004-iso-8601.md

```diff
@@ -0,0 +1,51 @@
+- Start Date: 2023-10-06
+- Contributors:
+  - [Zak Burke](zburke@ebsco.com)
+- RFC PRs:
+  - DRAFT REVIEW: https://github.com/folio-org/rfcs/pull/13
+  - PRELIMINARY REVIEW: TBD
+  - PUBLIC REVIEW: TBD
+  - FINAL REVIEW: TBD
+- Outcome: <!-- Leave this blank. Will eventually be either ACCEPTED or REJECTED -->
+
+# Date-time values must comply with IETF RFC-3339
+
+## Summary
+
+* Validation of date-time values must assert compliance with IETF RFC-3339.
+* Date-time values in storage must be compliant with IETF RFC-3339 Section 5.6.
```

Comment:
RFC-3339 describes how date time values can be represented over the internet, how does that apply to the internal storage a module uses?

Reply:
Section 5.6 is specifically about the format of those date-time values.  So I think it's valid, but I get the quibble.  Maybe just edit to

`Date-time values in storage must be compliant with the date/time format specified in IETF RFC-3339 Section 5.6.`

Reply:
What I'm missing is, why it is relevant for us to establish a policy on the storage format, when what is important is the external behaviour of the module?

For example, lets say I choose to store date / time as a numeric offset from 1st Jan 1970 UTC. AFAIK, that isn't compatible with representations defined in RFC-3339 however does allow the module to provide the intended behaviour of handling RFC-3339 formatted date / times from outside the module

Reply:
>  I don't think the TC (or this RFC) should restrict how the data is stored

Fine. I was thinking of API surfaces and storage as the same thing, but they don't have to be. Would this be acceptable if we say that dates must be presented and accepted _at API surfaces_ as RFC-3339 compliant values?

Reply:
Reviewing this, I would want to push for compliance with IETF RFC-3339 at both the API surface and the storage layer.
At the dev layer, consistency in storage means there's less surprise for a new person coming on when they run into a field in storage that is an exception. Maybe that's not a big deal when there's consistency with the developers and modules, but we know full well that individuals cycle off, some modules are orphaned, and discovering that the there's a field where date-time is stored differently than most can be a time-sink.
We have at least one reporting tool in the broader environment that pulls from the storage layer, so data as-stored is exposed to the user, and then the reporting user has the same problem as above.
Finally, if we're looking towards a future state where closely linked modules may directly access storage, that sets up another version of "people need to know this field isn't IETF RFC-3339" in a way that spans business logic in different modules.
I don't see an upside to leaving the storage representation to the implementer's discretion.

Reply:
> consistency in storage means there's less surprise for a new person coming on when they run into a field in storage that is an exception

A trade off with consistency is that it does not allow for context sensitive decisions and can impede progression of ideas

FOLIO has already actively chosen not to be consistent in storage when it chose to support both Spring and RMB based modules, as the Spring based modules tend to use regular SQL columns rather than the JSONB favoured by RMB

Given that, if consistency is our goal for modules, date-time storage would not be where I would start to try to address these concerns

> We have at least one reporting tool in the broader environment that pulls from the storage layer, so data as-stored is exposed to the user, and then the reporting user has the same problem as above.

That is a compromise the design of those tools chose to make by using the privileged position of system operators to access the internal storage of the modules

> Finally, if we're looking towards a future state where closely linked modules may directly access storage, that sets up another version of "people need to know this field isn't IETF RFC-3339" in a way that spans business logic in different modules

The FOLIO community seems really keen on using the shared database integration pattern. Given this has now come up in multiple conversations (also being included in the applications RFC), the sooner this is specifically discussed and the decision is made the better IMO

I find it odd that folks are concerned about how highly coupled the modules are and acknowledge that presents challenge for us, yet also want to chose an integration style that is known to increase the coupling between parts of a system

> I don't see an upside to leaving the storage representation to the implementer's discretion

My understanding of the format specified in RFC-3339 is that it is a text based format

Depending upon how strictly folks interpret the standard, setting the expectation that this representation is the only way date-time values can be stored would mean that all date time values in the DB would need to be stored as text

Compliance with that would mean that module developers would not be allowed to use the built in PostgreSQL [types](https://www.postgresql.org/docs/current/datatype-datetime.html) for date-time. Whilst some of those types are stated as compatible with RFC-3339, they are not limited to only that format and thus would not meet this policy

We already have modules that use those types e.g. [mod-data-import](https://github.com/folio-org/mod-data-import/blob/a84cf373a92201851ffad8b19a9a8fa3f9c58fe1/src/main/resources/liquibase/module/scripts/v-0.0.1/2023-08-28--00-00-create-queue-item-table.xml#L14C27-L14C27). Those modules would be invalidated if we chose this policy

Reply:
You're both making good points. My goal is for us to have a standard and that we follow it whenever it is reasonable to do so. If a module has some compelling reason to store `1697799665845` instead of `"2023-10-20T11:01:05.845Z"`, fine. But if a module wants to store `"2023-10-20T11:01:05"` instead of `"2023-10-20T11:01:05Z"` that strikes me as unfortunate. IOW, let's provide good defaults, expect them to be followed most of the time, and expect that folks can state a good reason when they don't follow it. Ignorance of the standard shall not constitute a good reason. 

If API surfaces are the only place where there is agreement about an applicable standard then I am fine limiting the scope of this RFC to API surfaces. If we can extend its scope to include that when date-time values are stored as strings they must also follow this standard then I think that's a win but it's not the part I'm most going to fight for.

Reply:
> If we can extend its scope to include that when date-time values are stored as strings they must also follow this standard then I think that's a win

I would be comfortable with that

Reply:
> if a module wants to store "2023-10-20T11:01:05" instead of "2023-10-20T11:01:05Z" that strikes me as unfortunate

I imagine that would also eventually expose itself in the API behaviour, unless the module consistently converted the date-time to UTC on input and inferred the UTC timezone for all values upon use

Reply:
@marcjohnson-kint, My goal is this: a developer should never interact with a date-time value that lacks a time zone. 

If the only place where there is agreement on this statement is where such values are represented as strings at API surfaces I am happy rewrite this RFC to further limit its scope. If we can establish corresponding expectations about storage in non-string formats (such as but not limited to PostgreSQL's built-in types) without getting mired in details, I think that would greatly increase the impact. 

Which do you prefer?

Reply:
> If the only place where there is agreement on this statement is where such values are represented as strings at API surfaces I am happy rewrite this RFC to further limit its scope

My opinion is that it would be better to limit the scope to the API representations / behaviour. I think it's obvious that there is not agreement on this

> If we can establish corresponding expectations about storage in non-string formats (such as but not limited to PostgreSQL's built-in types) without getting mired in details, I think that would greatly increase the impact

I'm biased against the standard being this strict and wide reaching

However, if a hard and consistent expectation that date time values MUST always have a timezone is what folks want to achieve then let's start that right at the beginning of this and consider extending the scope to cover the other areas where that might be relevant e.g. database storage, Kafka, S3 storage, file uploads, UI presentation

Reply:
Can you give an example of a situation where it is valuable to have a date-time value without a time zone?

Reply:
I don't have an example off the top of my head

The usual use case is when the application interprets the meaning based upon some other configuration e.g. I want to block out my calendar between 8-9 every morning irrespective of the time zone I'm currently in

Reply:
The "block my calendar" example you provide is an example of times unattached to dates and should be implemented as "between start date-time A and end date-time B, block my calendar from time X to time Y". i.e. it corresponds to times X and Y, not to date-times A and B. 

MODAUD-172 is about mod-audit failing to correctly handle a value that is compliant with RFC-3339, i.e. a date-time value that _has_ a timezone. It is a straight up bug: given unambiguous information, it returns the wrong result. Even if this RFC hand been written and accepted before MODAUD-1 were filed, it would still be a bug. 

My whole point in writing this RFC is that we can avoid a whole class of bugs related to making assumptions about date-times values without time zones if we do not accept date-time values without timezones. We can just choose not to have those problems. 

I understand the value of avoiding words like "must" that create rules that need exceptions be of the need to make context-sensitive decisions. There is real value in that. But an alternative problem is that without rules, people make context-ignorant decisions and that result in bugs and then push back against fixing the bugs because the guidelines (not "rules") say code should (not "must") act in a particular way, and besides the code is already in production and fixing it is expensive and other things probably rely on the implementation as-is so fixing the problem may actually cause _more problems_. All true, hence my very strong desire to do everything possible _not to have these problems in the first place_.

Reply:
All good points, @marcjohnson-kint, especially that if "rules get so diluted by exceptions that folks stop caring ... Ultimately, this comes down to what culture FOLIO wants to have". This is bang on target and puts words to an unspoken part of this RFC that (IMO) needs to be said out loud. Thank you for articulating it so clearly!

Personally, I trust a proposal more when it has withstood widespread attack than when it was been met with no objection. IOW, I greatly value the criticisms you have raised and regret that my pushback has come across as frustration.

Reply:
"Timestamp with time zone values" are fine. This RFC is aimed at avoiding the transmission or storage of date-time values _without_ an accompanying time zone. 

This RFC solves the problem of a time zone-less value being misinterpreted. Based on applicable standards, a date-time without a time zone should be interpreted as localtime. This could be the time zone configured by end-user's computer, or stored for the tenant, or of the server where Okapi is running, or the server where PostgreSQL is running, all of which will be wrong in at minimum 23/24 occurrences when a such a value is instead interpreted as UTC.